### PR TITLE
bugfix/java references

### PR DESCRIPTION
**Pull request by baywet**
- fixes a bug where java reference update requests would fail due to formatting
- updates test files

updates & related to:
https://github.com/microsoftgraph/msgraph-sdk-java/pull/576

### CODE REVIEW

File: src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs [hunk 1]
```diff
@@ -132,7 +133,53 @@ public static string TemplateName(this string templateFile)
...
+                                        .OrderBy(x => x.slashCount)
+                                        .Select(x => x.path)
+                                        .FirstOrDefault();
+            return string.IsNullOrEmpty(shortestCanonicalPath) ? string.Empty : lastIdReplacer.Replace(shortestCanonicalPath, string.Empty);
+        }
+        public static bool ShouldIncludeReferenceUpdateMethods(this OdcmObject c, OdcmModel m)
+        {
+            return GetCanonicalNavigationProperties(c, m).Any();
+        }
+        private static IEnumerable<OdcmProperty> GetCanonicalNavigationProperties(this OdcmObject c, OdcmModel m, bool includeContainsTarget = false, bool includeNonCollectionProperties = false)
```
Comment (zengin) [inline src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs:160]: Can you please add a summary for this method?

Comment (baywet) [reply src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs:184]: added.

File: test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java [hunk 2]
```diff
@@ -32,20 +32,27 @@
      * @param client         the service client
      * @param requestOptions the options for this request
      */
-    public UserReferenceRequest(@Nonnull final String requestUrl, @Nonnull final IBaseClient client, @Nullable final java.util.List<? extends com.microsoft.graph.options.Option> requestOptions) {
+    public UserReferenceRequest(String requestUrl, IBaseClient client, java.util.List<? extends com.microsoft.graph.options.Option> requestOptions) {
```
Comment (nikithauc) [inline test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java:35]: As the Nonnull annotation is removed, are null parameters allowed and if yes, are they handled by the code?

File: test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java [hunk 3]
```diff
@@ -32,20 +32,27 @@
...
         super(requestUrl, client, requestOptions, User.class);
     }
 
+    public void delete(final ICallback<? super User> callback) {
```
Comment (nikithauc) [inline test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java:39]: Can you please verify if the newly added methods  in this PR have summary?

File: test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java (hunk 2 above)
Comment (baywet) [reply test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java]: sorry about that, base rebase. Fixed

File: test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java (hunk 3 above)
Comment (baywet) [reply test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/UserReferenceRequest.java]: same thing, bad rebase.

File: test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/GroupCollectionRequest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import com.microsoft.graph.requests.extensions.GroupCollectionResponse;
+import com.microsoft.graph.requests.extensions.GroupCollectionRequestBuilder;
+import com.microsoft.graph.requests.extensions.GroupCollectionRequest;
+
+// **NOTE** This file was generated by a tool and any changes will be overwritten.
+
+/**
+ * The class for the Group Collection Request.
+ */
+public class GroupCollectionRequest extends BaseCollectionRequest<Group, GroupCollectionResponse, GroupCollectionPage> {
```
Comment (nikithauc) [inline test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/GroupCollectionRequest.java:29]: Is $count query option not applicable for  this class?

Comment (baywet) [reply test/Typewriter.Test/TestDataJava/com/microsoft/graph/requests/extensions/GroupCollectionRequest.java:29]: java currently doesn't have any method being generated for count...which is amazing to me that nobody complained about it so far!
it's captured here and will be tackled in a subsequent PR https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/issues/318

File: src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs [hunk 5]
```diff
@@ -157,7 +158,61 @@ public static string TemplateName(this string templateFile)
...
+        {
+            Func<OdcmProperty, bool> propertiesFilter = x => x.IsNavigation() && x.Projection.Type == c && (x.IsCollection || includeNonCollectionProperties) && (x.IsReference() || includeContainsTarget);
+            var entityContainerProperties = m.EntityContainer.Properties.Where(propertiesFilter);
+            var namespaceProperties = m.Namespaces
+                                            .SelectMany(x => x.Classes)
+                                            .SelectMany(x => x.Properties.Where(propertiesFilter));
+            return m.EntityContainer.Properties.OfType<OdcmSingleton>()
+                                        .SelectMany(x => x.Projection.Type.AsOdcmClass().Properties.Where(propertiesFilter))
+                                        .Union(entityContainerProperties.Any() ? entityContainerProperties : new List<OdcmProperty>())
+                                        .Union(namespaceProperties.Any() ? namespaceProperties : new List<OdcmProperty>());
```
Comment (zengin) [inline src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs:202]: not blocking, but I wonder what you think about this in terms of reading the logic faster:
```suggestion
            return m.EntityContainer.Properties
                .Union(m.Namespaces.SelectMany(x => x.Classes)
                    .SelectMany(x => x.Properties))
                .Union(m.EntityContainer.Properties.OfType<OdcmSingleton>()
                    .SelectMany(x => x.Projection.Type.AsOdcmClass().Properties))
                .Where(propertiesFilter);
```

**Review by zengin — APPROVED** [review]

**Review by MIchaelMainer — APPROVED** [review]

File: src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs (hunk 5 above)
Comment (baywet) [reply src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs]: thanks for the suggestion. After going back and forth I forgot to take a step back on this one.
I just pushed a new commit improving upon your suggestion, let me know what you think

**Review by zengin — APPROVED** [review]

**Pull request merged**